pkg/controller/statefulset/stateful_set_control_test.go

```diff
@@ -948,6 +954,219 @@ func TestStatefulSetControl_getSetRevisions(t *testing.T) {
 		})
 }
 
+func createPodsAndRevisionForSet(set *apps.StatefulSet, revision int64) (retRev *apps.ControllerRevision, retPods []*v1.Pod) {
+	retRev = newRevisionOrDie(set, revision)
+	replicaCount := int(*set.Spec.Replicas)
+	for i := 0; i < replicaCount; i++ {
+		pod := newStatefulSetPod(set, i)
+		setPodRevision(pod, retRev.Name)
+		retPods = append(retPods, pod)
+	}
+	return
+}
+
+func newStatefulSetOnDelete() *apps.StatefulSet {
+	set := newStatefulSet(3)
+	set.Spec.UpdateStrategy = apps.StatefulSetUpdateStrategy{
+		Type: apps.OnDeleteStatefulSetStrategyType,
+	}
+	return set
+}
+
+func runForOnDeletePodScenarios(t *testing.T,
+	testFn func(*testing.T, *apps.ControllerRevision, *apps.StatefulSet, []*v1.Pod, []*apps.ControllerRevision),
+) {
+
+	type SetPodRevision struct {
+		set      *apps.StatefulSet
+		pods     []*v1.Pod
+		revision *apps.ControllerRevision
+	}
+
+	setNextRevision := func(setPodRevisionByRevisionNumber []SetPodRevision, revisionNumber int, currRev *apps.ControllerRevision, changeImage string) []SetPodRevision {
+		set := newStatefulSetOnDelete()
+		rev := &apps.ControllerRevision{}
+		var pods []*v1.Pod
+		if revisionNumber != 0 {
+			set.Spec.Template.Spec.Containers[0].Image = changeImage
+			set.Status.CurrentRevision = currRev.Name
+			set.Status.CollisionCount = new(int32)
+			rev, pods = createPodsAndRevisionForSet(set, int64(revisionNumber)+1)
+			set.Status.UpdateRevision = rev.Name
+		} else {
+			set.Status.CollisionCount = new(int32)
+			rev, pods = createPodsAndRevisionForSet(set, int64(revisionNumber)+1)
+			set.Status.CurrentRevision = rev.Name
+			set.Status.UpdateRevision = rev.Name
+		}
+		setPodRevisionByRevisionNumber = append(setPodRevisionByRevisionNumber, SetPodRevision{
+			set:      set,
+			pods:     pods,
+			revision: rev,
+		})
+		return setPodRevisionByRevisionNumber
+	}
+
+	setPodRevisionByRevisionNumber := []SetPodRevision{}
+
+	buildSetPodRevisionByRevisionNumber := func(setPodRevisionByRevisionNumber []SetPodRevision) []SetPodRevision {
+		setPodRevisionByRevisionNumber = setNextRevision(setPodRevisionByRevisionNumber, 0, nil, "")
+		setPodRevisionByRevisionNumber = setNextRevision(setPodRevisionByRevisionNumber, 1, setPodRevisionByRevisionNumber[0].revision, "foo")
+		setPodRevisionByRevisionNumber = setNextRevision(setPodRevisionByRevisionNumber, 2, setPodRevisionByRevisionNumber[0].revision, "bar")
+		return setPodRevisionByRevisionNumber
+	}
+
+	setPodRevisionByRevisionNumber = buildSetPodRevisionByRevisionNumber(setPodRevisionByRevisionNumber)
+
+	tests := []struct {
+		name      string
+		expected  *apps.ControllerRevision
+		set       *apps.StatefulSet
+		pods      []*v1.Pod
+		revisions []*apps.ControllerRevision
+	}{
+		{
+			// Test case: stable statefulset
+			name:      "StableRevision",
+			expected:  setPodRevisionByRevisionNumber[0].revision,
+			set:       setPodRevisionByRevisionNumber[0].set,
```

Comment:
I wanted to do that (avoid reusing or keeping state), specially when you mentioned builder funcs on previous reviews of the previous iteration of the other PR. But since each test tests something that "walked through a list of steps" that were basically the steps of the previous tests, it seems weird to recreate everything from scratch on each test.

If we look at tests that are already accepted (on master) they do the same, when testing stuff regarding revision. (But they don't use table driven tests...)